Assets/Mirage/Components/Visibility/SpatialHash/SpatialHashSystem.cs

```diff
@@ -0,0 +1,264 @@
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void AddObject(int i, int j, T obj)
+            {
+                int index = i + j * Width;
+                if (Points[index].objects == null)
+                {
+                    Points[index].objects = new HashSet<T>();
+                }
+
+                Points[index].objects.Add(obj);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public HashSet<T> GetObjects(int i, int j)
+            {
+#if DEBUG
+                if (i < 0) throw new IndexOutOfRangeException($"i ({i}) is less than zero");
+                if (j < 0) throw new IndexOutOfRangeException($"j ({j}) is less than zero");
+                if (i >= Width) throw new IndexOutOfRangeException($"i ({i}) is greater than {Width}");
+                if (j >= Height) throw new IndexOutOfRangeException($"j ({j}) is greater than {Height}");
+#endif
+                return Points[i + j * Width].objects;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void CreateSet(int i, int j)
+            {
+                Points[i + j * Width].objects = new HashSet<T>();
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool InBounds(Vector2 position)
+            {
+                return (-Extents.x <= position.x && position.x <= Extents.x)
+                    && (-Extents.y <= position.y && position.y <= Extents.y);
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool InBounds(int x, int y)
+            {
+                // inclusive lower bound
+                return (0 <= x && x < Width)
+                    && (0 <= y && y < Height);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool IsVisible(Vector2 target, Vector2 player, int range)
+            {
+                // if either is out of bounds, not visible
+                if (!InBounds(target) || !InBounds(player)) return false;
+
+                ToGridIndex(target, out int xt, out int yt);
+                ToGridIndex(player, out int xp, out int yp);
+
+                return AreClose(xt, xp, range) && AreClose(yt, yp, range);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            bool AreClose(int a, int b, int range)
+            {
+                int min = a - range;
+                int max = a + range;
+
+                return max <= b && b <= min;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void ToGridIndex(Vector2 position, out int x, out int y)
+            {
+                float fx = position.x - Offset.x;
+                float fy = position.y - Offset.y;
+
+                x = Mathf.RoundToInt(fx / GridSize.x);
+                y = Mathf.RoundToInt(fy / GridSize.y);
```

Comment:
Currently bugged near edge, might need extra offset of  `GridSize/2`  or something.

Unit tests would be good for this function.